### PR TITLE
Hook up RAWR data fetcher

### PR DESCRIPTION
**Pull request by zerebubuth**
Added the ability to load RAWR tiles for processing from S3, or generate them directly from the database. The latter is useful for testing, and bypasses the round-trip via the msgpack-encoded `.zip` file, but is otherwise identical (and takes a while).

Added a `DataFetcher` implementation which is able to dynamically switch between database and RAWR tiles based on the zoom level(s) of the incoming tiles.

Plenty of TODOs in here to follow up, but hopefully this is a "save point" and we can clear the TODOs up afterwards.

### CODE REVIEW

File: tilequeue/worker.py [hunk 1]
```diff
@@ -189,59 +220,69 @@ def __call__(self, stop):
...
                 continue
-            if data is None:
+            if all_data is None:
                 saw_sentinel = True
                 break
 
-            coord = data['coord']
-            nominal_zoom = coord.zoom + self.metatile_zoom
-            unpadded_bounds = coord_to_mercator_bounds(coord)
+            for fetch, data in self.fetcher.start(all_data):
```
Comment (rmarianski) [inline tilequeue/worker.py]: Is `start` the best name for it? Seems like it's more of a `group` or `parcel` kind of operation. Names are hard :)

File: tilequeue/query/__init__.py [hunk 2]
```diff
@@ -1,9 +1,180 @@
...
+    outputs = set()
+    for value in inputs:
+        if value.startswith('Multi'):
+            value = value[len('Multi'):]
+        outputs.add(value.lower())
+
+    if outputs:
+        return outputs
+    else:
+        return None
```
[the remark this answers is not in the document]
Comment (zerebubuth) [reply tilequeue/query/__init__.py]: Thanks, that's a more succinct way of writing it. Fixed in 57a727d.

File: tilequeue/query/__init__.py [hunk 3]
```diff
@@ -1,9 +1,180 @@
...
+
+def _parse_yaml_functions(process_yaml_cfg):
+    from vectordatasource.meta.python import make_function_name_props
+    from vectordatasource.meta.python import make_function_name_min_zoom
+    from vectordatasource.meta.python import output_kind
+    from vectordatasource.meta.python import output_min_zoom
+    import os.path
+
+    # can't handle "callable" type - how do we get the min zoom fn?
+    assert process_yaml_cfg['type'] == 'parse'
```
[the remark this answers is not in the document]
Comment (zerebubuth) [reply tilequeue/query/__init__.py]: Sure! There's a small difference that the output calc function mapping isn't available for min zooms (yet?), but I've moved all the relevant logic to `command.py` so we can merge it later if we grow that functionality. (57a727d).

File: tilequeue/rawr.py [hunk 4]
```diff
@@ -308,6 +309,23 @@ def make_rawr_zip_payload(rawr_tile, date_time=None):
     return buf.getvalue()
 
 
+def unpack_rawr_zip_payload(io):
+    """unpack a zipfile and turn it into a callable "tables" object."""
+    # TODO: the io we get from S3 is streaming, so we can't seek on it, but
+    # zipfile seems to require that. should we write a temporary file rather
+    # than storing it all in memory?
+    buf = io.BytesIO(io.read())
```
[the remark this answers is not in the document]
Comment (zerebubuth) [reply tilequeue/rawr.py]: Yup, cool. Rewrote the comment in 57a727d to explain that's what's happening.

File: tilequeue/rawr.py [hunk 5]
```diff
@@ -339,3 +357,41 @@ def __call__(self, rawr_tile):
...
+                Bucket=self.bucket,
+                Key=location,
+        )
+
+        return response
+
+    def __call__(self, tile):
+        # TODO: i guess this throws an exception if the object is missing, or
+        # there's a server error. do we want to catch it, or is propagating it
+        # the right option?
```
[the remark this answers is not in the document]
Comment (zerebubuth) [reply tilequeue/rawr.py]: You're right - I think we'll want to have a "partial build" in dev, and we probably don't want it to be failing to build anything outside of that. Eventually, I think we'll want a (shared) global build for dev and prod, but until then we might want a partial one. I added the ability to optionally ignore 404 errors, translating the exception into an empty tile, in 57a727d.

File: tilequeue/rawr.py [hunk 6]
```diff
@@ -339,3 +357,41 @@ def __call__(self, rawr_tile):
...
+        if self.io_pool:
+            response = self.thread_pool.apply(
+                self._get_object, (tile,))
+        else:
+            response = self._get_object(tile)
+
+        # check that the response isn't a delete marker.
+        assert not response['DeleteMarker']
```
[the remark this answers is not in the document]
Comment (zerebubuth) [reply tilequeue/rawr.py]: I was following a bit of example code, so it's possible that we don't ever encounter objects with delete markers on them in our stack.

File: tilequeue/worker.py (hunk 1 above)
Comment (zerebubuth) [reply tilequeue/worker.py]: You're right, `start` is a pretty awful name for it. Perhaps it fitted better back when it was a `@contextmanager`, but now it's `yield`ing a sequence of `(fetch, data)` tuples then it's not an intuitive name for it.

Would something that indicates iteration be better? Like `items()`, `each()` or `fetch_tiles()`?

Comment (rmarianski) [reply tilequeue/worker.py]: Yup, I like any of those names.

File: tilequeue/rawr.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rmarianski) [reply tilequeue/rawr.py]: Yea, I suppose we'll never practically see it. When we delete via lifecycle rules, or I suppose any deletes really, we should  be reading from a new prefix anyway.

File: tilequeue/worker.py (hunk 1 above)
Comment (zerebubuth) [reply tilequeue/worker.py]: Cool, thanks. Fixed in 5840e9a.

**Pull request merged**